conwayste/src/ui/context.rs

```diff
@@ -0,0 +1,216 @@
+    /// };
+    /// my_widget.on(context::EventType::Click, Box::new(handler));
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// * It is an error to call this from within a handler.
+    fn on(&mut self, what: EventType, f: Handler) -> Result<(), Box<dyn Error>>;
+
+    /// Emit an event -- call all handlers for this event's type (as long as they return NotHandled)
+    ///
+    /// # Errors
+    ///
+    /// * It is an error to call this from within a handler.
+    /// * The first error to be returned by a handler will be returned here, and no other handlers
+    ///   will run.
+    fn emit(&mut self, event: &Event, uictx: &mut UIContext) -> Result<(), Box<dyn Error>>;
+}
+
+impl_downcast!(EmitEvent);
+
+/// Implement EmitEvent for a widget (though strictly speaking non-widgets can implement it).
+///
+/// # Example
+///
+/// ```
+/// struct MyWidget {
+///     handlers: Option<HandlerMap>,
+///     ...
+/// }
+///
+/// impl MyWidget {
+///     fn new() -> Self {
+///         MyWidget {
+///             handlers: Some(context::HandlerMap::new()),
+///             ...
+///         }
+///     }
+/// }
+/// // top level of the module
+/// impl_emit_event!(MyWidget, self.handlers);
+/// ```
+#[macro_export]
+macro_rules! impl_emit_event {
+    ($widget_name:ty, self.$handler_field:ident) => {
+        impl crate::ui::context::EmitEvent for $widget_name {
+            /// Setup a handler for an event type
+            fn on(&mut self, what: crate::ui::context::EventType, hdlr: crate::ui::context::Handler) -> Result<(), Box<dyn std::error::Error>> {
+                let handlers = self.$handler_field
+                    .as_mut()
+                    .ok_or_else(|| -> Box<dyn std::error::Error> {
+                        format!("during .on({:?}, ...): a .emit call is in progress for {} widget", what, stringify!($widget_name)).into()
+                    })?;
+
+                let handler_vec: &mut Vec<crate::ui::context::Handler>;
+                if let Some(vref) = handlers.get_mut(&what) {
+                    handler_vec = vref;
+                } else {
+                    handlers.insert(what, vec![]);
+                    handler_vec = handlers.get_mut(&what).unwrap();
+                }
+                handler_vec.push(hdlr);
+                Ok(())
+            }
+
+            /// Emit an event -- call all handlers for this event's type (as long as they return NotHandled)
+            fn emit(&mut self, event: &crate::ui::context::Event, uictx: &mut crate::ui::context::UIContext) -> Result<(), Box<dyn std::error::Error>> {
+                use crate::ui::context::Handled::*;
+                // HACK: prevent a borrow error when calling handlers
+                let mut handlers = self.$handler_field
+                    .take()
+                    .ok_or_else(|| -> Box<dyn std::error::Error> {
+                        format!("during .on({:?}, ...): a .emit call is in progress for {} widget",
```

Comment:
Is the error message accurate here? It seems to match the one in `on(...)`, but we're actually in `emit(...)`. Just double checking.

Reply:
Good catch. I'll take a closer look. I think my logic at the time was that if the handler field is `None`, either a `.on` (register handler) or a `.emit` (call handler) was already in progress for this widget.